### ISSUE TITLE
Improve custom resources integration

### ISSUE THREAD

**Issue by medikoo** [serverless/serverless#6575]
Currently after all resources are generated, custom ones (defined at `resources.Resources) are deeply merged to it:
https://github.com/serverless/serverless/blob/66ef74b0c6f09d9435921155d8820cbc3cd77432/lib/plugins/aws/package/lib/mergeCustomProviderResources.js#L15-L18

It's unsafe operation, as it unconditionally merges all arrays and objects, producing broken configurations in some scenarios (e.g. merging two independent `Fn::Join`instructins, as reported here #6569). Also it doesn't apply any validation so technically two resources of very different types, if they share logical id, will be merged silently.

--- 
Proposal for new approach

Distinct between newly introduced custom resources, and extensions to ones generated by Fremework.

1. Use flat (`Object.assign` like) merge for `resources.Resources`, but  error if given resource was already generated by the framework (do not silently override)

2. Introduce `resources.extensions` for definition of resource extensions which should be applied to resources already generated by the framework. In this case it's `Properties` that should be expected top level, and they're merged flat way (so if framework generated given property, it'll be overriden by custom resource completely)

**Comment by glb** [comment]
> Accept newly introduced user resources (not extensions) at resources.additions. Use direct (Object.assign like) assignment to apply them on Resources, but error if given resource was already generated by the framework (do not silently override)

@medikoo Do you have an intended schema for `resources.additions`? Is it like this?

```json
{
    additions: {
        oneOf: [{
            type: 'object',
            patternProperties: {
                '^[a-zA-Z0-9]{1,255}$': {
                    type: 'object',
                }
            }
        }, {
            type: 'array',
            items: {
                type: 'object',
                patternProperties: {
                    '^[a-zA-Z0-9]{1,255}$': {
                        type: 'object',
                    }
                }
            }
        }]
    }
}
```

and then something that copies from `resources.additions` into `Resources` with failures for duplicate keys? I think that the null check would happen through schema validation, yes?

**Comment by medikoo** [comment]
> @medikoo Do you have an intended schema for resources.additions? Is it like this?

This is not yet implemented, so it's out of scope (Schema should be added together with implementation).

It's only `resources.extensions` which is already in, and which misses the schema.

**Comment by glb** [comment]
@medikoo the reason I asked if you had an intended schema was because I am thinking of implementing `resources.additions` and wanted to know if you had a direction in mind for the schema before I started.

**Comment by medikoo** [comment]
> @medikoo the reason I asked if you had an intended schema was because I am thinking of implementing resources.additions and wanted to know if you had a direction in mind for the schema before I started.

@glb great to hear that! PR on that would be very welcome. I would follow same pattern as in `extensions`. I believe we should not support there an _array_ form. Why did you envision such type there?

**Comment by glb** [comment]
@medikoo our team has found the array form of `resources` to be extremely useful -- they can split up resources into multiple files based on what makes sense to be grouped together and then include them using:

```yaml
resources:
- ${file(file-1.yml)}
- ${file(file-2.yml)}
- ${file(file-3.yml)}
```

I would want to maintain that functionality (yes I know we would need to remove the `Resources` at the top level of the files):

```yaml
resources:
  additions:
  - ${file(file-1.yml)}
  - ${file(file-2.yml)}
  - ${file(file-3.yml)}
```

**Comment by medikoo** [comment]
> @medikoo our team has found the array form of resources to be extremely useful 

@glb that's understood, and I agree with that. We have requests for many places where users would like to have such format available. In my eyes addressing those _reported_ places one by one is not viable so I was thinking about providing a solution that will allow to use that format in any object configuration. Have you checked my proposal here: https://github.com/serverless/serverless/issues/8037#issuecomment-674751706 ?

Now, concerning schema validation, note that this validation is applied on already normalized and resolved config, so _after_ such variable arrays are resolved (onto an object).

**Comment by glb** [comment]
@medikoo I have seen your proposal in #8037 ... I'm not sure where to start with that, whereas I think I could possibly implement what I've described here. I was thinking that I would try to avoid the DX issue I reported in #8118 by leaving the array as long as possible (ideally until after validation, which would catch the problem and report it quite nicely, I think), but I haven't started implementing yet so I don't know how difficult that would be.

**Comment by medikoo** [comment]
> #8118 by leaving the array as long as possible

@glb thanks for proposal. Still when implementing the config validation, we agreed to validate fully resolved config, after all variables are being resolved, and array notation is kind of part of variable notations (arrays are allowed in few specific places purely to enable to rely on few sources as resolved by variables).

Trying to shift validation somewhere in a middle makes things more complicated and obscure. It'll definitely make things way more confusing for most of us and will increase a maintenance cost.

Also I believe that what I proposed provides a very clean solution for your case. Can you explain why you feel it's not the right approach?

**Comment by glb** [comment]
@medikoo the thing that I'm most concerned about is the experience for the user of the framework. You have a vision in mind for how the issue in #8118 can be properly resolved to tell the user which element of their array is incorrect; I don't exactly understand how it works but will defer to your superior knowledge of the internals. If the user experience is functionally the same as "the third element of `resources.additions` should be an object but was null" then that's fantastic.

I don't currently have the deep knowledge of the framework to attempt #8037 and `resources.additions` is not something I can use without having the array option, so I'm going to hold off on attempting to implement #6575. If someone else has a need or desire to take this on, please go for it!

**Comment by medikoo** [comment]
>  If the user experience is functionally the same as "the third element of resources.additions should be an object but was null" then that's fantastic.

Yes, intention is to have exactly like that (or very similar) error.

Indeed #8037 is less trivial, I've added _How to implement it?_ proposal here in issue description

**Comment by rkj1980** [comment]
Sent from ProtonMail mobile

-------- Original Message --------
